src/hooks/useOutsideClick.ts

```diff
@@ -8,32 +8,48 @@ import { useEffect, useRef, useCallback } from "react";
  * @param ref Ref whose outside click needs to be listened to
  * @param handler Callback to fire on outside click
  * @param when A boolean which which activates the hook only when it is true. Useful for conditionally enable the outside click
+ * @param detectIframe A boolean which enables detecting click on outside iframe
+ * # caveats for iframe
+ * will only detect the first click on iframe, need to focus back to window, and click on iframe to detect again
+ * don't call alert on handler when using iframe
+ * focus to iframe via tab navigation also triggers window.blur
  */
 function useOutsideClick(
   ref: MutableRefObject<HTMLElement | null>,
-  handler: (e: MouseEvent) => any,
-  when: boolean = true
+  handler: (event: FocusEvent | MouseEvent) => any,
+  when: boolean = true,
+  detectIframe: boolean = true
 ): void {
   const savedHandler = useRef(handler);
 
-  const memoizedCallback = useCallback((e: MouseEvent) => {
-    if (ref && ref.current && !ref.current.contains(e.target as Element)) {
-      savedHandler.current(e);
+  const memoizedCallback = useCallback((event: MouseEvent) => {
+    if (ref?.current && !ref.current.contains(event.target as Element)) {
+      savedHandler.current(event);
     }
   }, []);
 
-  useEffect(() => {
-    savedHandler.current = handler;
-  });
+  const memoizedOnBlurCallback = useCallback((event: FocusEvent) => {
+    // Note: on firefox clicking on iframe triggers blur, but only on
+    // next event loop it becomes document.activeElement
+    // https://stackoverflow.com/q/2381336#comment61192398_23231136
+    setTimeout(() => {
```

Comment:
Can we use `useTimeout` here?

Reply:
using `useTimeout` here gives this error:
```
Error: Uncaught [Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
    1. You might have mismatching versions of React and the renderer (such as React DOM)
    2. You might be breaking the Rules of Hooks
    3. You might have more than one copy of React in the same app
```

Reply:
Could you paste your snippet here?  @qiweiii

Reply:
@imbhargav5 I found out Iwas using timeout incorectly previously, but still `useTimeout` is not suitable here because I have to declare `useTimeout`'s callback outside of this useCallback, and as `useTimeout` does not take aruguments, I cannot pass `event` into the timeout callback